### PR TITLE
Sensor events splitup

### PR DESCRIPTION
**Pull request by gast04**
Split up `sensorEvents` into one message per sensor, and immediately publish once the measurement is ready.

- [ ] update xx
- [x] update provisioning test

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Can you rebase now that the interrupts PR has been merged?

**Comment by gast04** [comment]
wilco

File: selfdrive/locationd/locationd.cc [hunk 1]
```diff
@@ -521,11 +522,25 @@ int Localizer::locationd_thread() {
     }
 
     // 100Hz publish for notcars, 20Hz for cars
-    const char* trigger_msg = sm["carParams"].getCarParams().getNotCar() ? "sensorEvents" : "cameraOdometry";
-    if (sm.updated(trigger_msg)) {
+    bool kalman_msg = false;
```
Comment (adeebshihadeh) [inline selfdrive/locationd/locationd.cc:525]: let’s keep the trigger msg logic here the same and just pick either the gyro or accelerometer to replace sensorEvents

File: selfdrive/locationd/locationd.cc [hunk 2]
```diff
@@ -521,11 +522,25 @@ int Localizer::locationd_thread() {
...
+        sm.updated("gyroscope") && sm.updated("magnetometer"))
+    {
+      kalman_msg = true;
+    }
+
+    if (kalman_msg) {
       bool inputsOK = sm.allAliveAndValid();
-      bool sensorsOK = sm.alive("sensorEvents") && sm.valid("sensorEvents");
       bool gpsOK = this->isGpsOK();
+      bool sensorsOK = true;
```
Comment (adeebshihadeh) [inline selfdrive/locationd/locationd.cc:540]: use SubMaster::allAliveAndValid https://github.com/commaai/cereal/blob/3baa20e1da5d88dcb1d3ae9678471eb8013958f2/messaging/messaging.h#L75

File: selfdrive/sensord/tests/test_sensord.py [hunk 3]
```diff
@@ -79,30 +80,49 @@
...
+  events = {}
+  for stype in sensor_types:
+    esocks[stype] = messaging.sub_sock(stype, timeout=0.1)
+    events[stype] = []
+
+  start_time_sec = time.monotonic()
+  while time.monotonic() - start_time_sec < duration_sec:
+    for esock in esocks:
+      events[esock] += messaging.drain_sock(esocks[esock])
+    time.sleep(0.01)
```
Comment (adeebshihadeh) [inline selfdrive/sensord/tests/test_sensord.py:116]: this sleep can be longer, the buffers are quite big
```suggestion
    time.sleep(0.1)
```

File: selfdrive/sensord/tests/test_sensord.py [hunk 4]
```diff
@@ -79,30 +80,49 @@
...
+    # to check one is enough, fails on if bmx is not present
+    with SMBus(SENSOR_BUS, force=True) as bus:
+      bus.read_byte_data(I2C_ADDR_BMX_ACCEL, BMX_REG_ID)
+    return True
+  except Exception:
+    return False
+
+def read_sensor_events(sensor_types, duration_sec):
+  esocks = {}
+  events = {}
```
Comment (adeebshihadeh) [inline selfdrive/sensord/tests/test_sensord.py:107]: use a defaultdict, then you don’t need to do the bmx check later

File: selfdrive/test/process_replay/regen.py [hunk 5]
```diff
@@ -247,7 +296,8 @@ def regen_segment(lr, frs=None, outdir=FAKEDATA, disable_tqdm=False):
       # ensure all procs are running
       for d, procs in fake_daemons.items():
         for p in procs:
-          if not p.is_alive():
+          if not p.is_alive() and d != 'sensord':
+            # not all sensors must have events (temperatureSensor, bmx sockets)
```
Comment (adeebshihadeh) [inline selfdrive/test/process_replay/regen.py:300]: let’s just not replay the bmx msgs at all in regen

File: selfdrive/sensord/sensors_qcom2.cc [hunk 6]
```diff
@@ -148,33 +124,31 @@ int sensor_loop() {
...
   }
 
   if (!has_magnetometer) {
     LOGE("No magnetometer present");
-    delete i2c_bus_imu;
     return -1;
   }
 
-  PubMaster pm({"sensorEvents"});
+  PubMaster pm({PM_GYRO, PM_GYRO_2, PM_ACCEL, PM_ACCEL_2, PM_TEMP, PM_LIGHT, PM_MAGN});
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors_qcom2.cc:151]: no need to define the socket names as constants

File: selfdrive/test/process_replay/regen.py (hunk 5 above)
Comment (gast04) [reply selfdrive/test/process_replay/regen.py:300]: ok, but this still holds for the temperature sensor, some routes dont have temperature events,
dont replay temp too?

File: selfdrive/sensord/sensors_qcom2.cc (hunk 6 above)
Comment (gast04) [reply selfdrive/sensord/sensors_qcom2.cc:151]: thought its more nice this way, cause we use this constants in the sensors too^^

File: selfdrive/test/process_replay/regen.py (hunk 5 above)
Comment (adeebshihadeh) [reply selfdrive/test/process_replay/regen.py:300]: yeah that's fine. they're not used anywhere right now

File: selfdrive/locationd/test/test_locationd.py [hunk 7]
```diff
@@ -14,7 +14,8 @@
 
 class TestLocationdProc(unittest.TestCase):
   MAX_WAITS = 1000
-  LLD_MSGS = ['gpsLocationExternal', 'cameraOdometry', 'carState', 'sensorEvents', 'liveCalibration']
+  LLD_MSGS = ['gpsLocationExternal', 'cameraOdometry', 'carState', 'liveCalibration',
+    'accelerometer', 'gyroscope', 'magnetometer', 'lightSensor', 'temperatureSensor']
```
[the remark this answers is not in the document]
Comment (gast04) [reply selfdrive/locationd/test/test_locationd.py]: done

File: selfdrive/locationd/locationd.cc (hunk 1 above)
Comment (gast04) [reply selfdrive/locationd/locationd.cc]: done

File: selfdrive/locationd/locationd.cc (hunk 2 above)
Comment (gast04) [reply selfdrive/locationd/locationd.cc]: done

File: selfdrive/sensord/tests/test_sensord.py (hunk 3 above)
Comment (gast04) [reply selfdrive/sensord/tests/test_sensord.py]: done

File: selfdrive/sensord/tests/test_sensord.py (hunk 4 above)
Comment (gast04) [reply selfdrive/sensord/tests/test_sensord.py]: done

File: selfdrive/sensord/sensors_qcom2.cc (hunk 6 above)
Comment (gast04) [reply selfdrive/sensord/sensors_qcom2.cc]: done

File: selfdrive/sensord/sensors_qcom2.cc [hunk 8]
```diff
@@ -67,37 +69,19 @@ void interrupt_loop(int fd, std::vector<Sensor *>& sensors, PubMaster& pm) {
...
-    std::vector<capnp::Orphan<cereal::SensorEventData>> collected_events;
-    collected_events.reserve(sensors.size());
-
     for (Sensor *sensor : sensors) {
-      auto orphan = orphanage.newOrphan<cereal::SensorEventData>();
-      auto event = orphan.get();
-      if (!sensor->get_event(event)) {
+      MessageBuilder msg;
+      std::string service;
+      if (!sensor->get_event(msg, service, ts)) {
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors_qcom2.cc:75]: the sensors classes shouldn't know anything about services. I'd make this an std::map, something like:
```
std::map sensors = {
  lsmGyroClass, "gyroscope",
  bmxGyroClass, "gyroscope2",
};
```

File: selfdrive/sensord/sensors/bmx055_accel.cc [hunk 9]
```diff
@@ -6,7 +6,7 @@
 #include "common/timing.h"
 #include "common/util.h"
 
-BMX055_Accel::BMX055_Accel(I2CBus *bus) : I2CSensor(bus) {}
+BMX055_Accel::BMX055_Accel(I2CBus *bus, uint64_t init_delay) : I2CSensor(bus, init_delay) {}
```
[the remark this answers is not in the document]
Comment (gast04) [reply selfdrive/sensord/sensors/bmx055_accel.cc:9]: hmm, mean we need to init for every sensor anyway, but yeah we can do like:
```
  virtual bool is_data_valid(uint64_t st, uint64_t ct) {
    return (ct - st) > init_delay;
  }
```
in sensor.h then all the other changes could be removed, and just initialize the value after the sensor creation

File: selfdrive/sensord/sensors_qcom2.cc (hunk 8 above)
Comment (gast04) [reply selfdrive/sensord/sensors_qcom2.cc:75]: yea, agree on that, thought about that as well, but thought then everything is more contained in the sensor

File: selfdrive/sensord/tests/test_sensord.py [hunk 10]
```diff
@@ -79,30 +79,35 @@
 }
 
 SENSOR_BUS = 1
-I2C_ADDR_LSM = 0x6A
+I2C_ADDR_BMX_ACCEL = 0x18
+BMX_REG_ID = 0x00
```
Comment (adeebshihadeh) [inline selfdrive/sensord/tests/test_sensord.py:83]: these are unused

File: selfdrive/sensord/sensors_qcom2.cc [hunk 11]
```diff
@@ -27,17 +28,22 @@
...
-
-// filter first values (0.5sec) as those may contain inaccuracies
 uint64_t init_ts = 0;
-constexpr uint64_t init_delay = 500*1e6;
 
-void interrupt_loop(int fd, std::vector<Sensor *>& sensors, PubMaster& pm) {
+void send_message(PubMaster& pm, MessageBuilder& msg, std::string &service) {
+  std::lock_guard<std::mutex> lock(pm_mutex);
+  pm.send(service.c_str(), msg);
+}
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors_qcom2.cc:36]: we should just use separate PubMasters for each loop

File: selfdrive/sensord/sensors_qcom2.cc [hunk 12]
```diff
@@ -179,33 +169,30 @@ int sensor_loop() {
...
+      if (!sensor->get_event(msg)) {
+        continue;
+      }
 
-    if (nanos_since_boot() - init_ts < init_delay) {
-      continue;
-    }
+      if (!sensor->is_data_valid(init_ts, nanos_since_boot())) {
+        continue;
+      }
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors_qcom2.cc:193]: check before getting the event

File: selfdrive/sensord/sensors/bmx055_gyro.cc [hunk 13]
```diff
@@ -84,6 +84,7 @@ bool BMX055_Gyro::get_event(cereal::SensorEventData::Builder &event) {
   float y = -DEG2RAD(read_16_bit(buffer[2], buffer[3]) * scale);
   float z = DEG2RAD(read_16_bit(buffer[4], buffer[5]) * scale);
 
+  auto event = msg.initEvent().initGyroscope();
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors/bmx055_gyro.cc:87]: shouldn't this be gyro2?

File: selfdrive/sensord/sensors/bmx055_accel.cc [hunk 14]
```diff
@@ -73,6 +74,7 @@ bool BMX055_Accel::get_event(cereal::SensorEventData::Builder &event) {
   float y = -read_12_bit(buffer[2], buffer[3]) * scale;
   float z = read_12_bit(buffer[4], buffer[5]) * scale;
 
+  auto event = msg.initEvent().initAccelerometer();
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors/bmx055_accel.cc:77]: same here

File: selfdrive/sensord/sensors/lsm6ds3_accel.cc [hunk 15]
```diff
@@ -113,10 +114,14 @@ bool LSM6DS3_Accel::get_event(cereal::SensorEventData::Builder &event) {
   float y = read_16_bit(buffer[2], buffer[3]) * scale;
   float z = read_16_bit(buffer[4], buffer[5]) * scale;
 
+  auto event = msg.initEvent().initAccelerometer();
   event.setSource(source);
   event.setVersion(1);
   event.setSensor(SENSOR_ACCELEROMETER);
   event.setType(SENSOR_TYPE_ACCELEROMETER);
+  if (ts != 0) {
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors/lsm6ds3_accel.cc:122]: is this check necessary? the timestamp should default to 0 anyway

File: selfdrive/sensord/sensors/lsm6ds3_gyro.cc [hunk 16]
```diff
@@ -116,10 +117,14 @@ bool LSM6DS3_Gyro::get_event(cereal::SensorEventData::Builder &event) {
   float y = DEG2RAD(read_16_bit(buffer[2], buffer[3]) * scale);
   float z = DEG2RAD(read_16_bit(buffer[4], buffer[5]) * scale);
 
+  auto event = msg.initEvent().initGyroscope();
   event.setSource(source);
   event.setVersion(2);
   event.setSensor(SENSOR_GYRO_UNCALIBRATED);
   event.setType(SENSOR_TYPE_GYROSCOPE_UNCALIBRATED);
+  if (ts != 0) {
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors/lsm6ds3_gyro.cc:125]: same here

File: selfdrive/sensord/sensors/bmx055_gyro.cc (hunk 13 above)
Comment (gast04) [reply selfdrive/sensord/sensors/bmx055_gyro.cc:87]: oh boy, good catch

File: selfdrive/sensord/sensors/lsm6ds3_accel.cc (hunk 15 above)
Comment (gast04) [reply selfdrive/sensord/sensors/lsm6ds3_accel.cc]: we could remove the `has_interrupt_enabled` if too, lsm will always run on interrupt

File: selfdrive/sensord/sensors_qcom2.cc (hunk 12 above)
Comment (gast04) [reply selfdrive/sensord/sensors_qcom2.cc:190]: I think it makes sense to do it after, just to have the full communication with the sensor

File: selfdrive/sensord/sensors_qcom2.cc (hunk 11 above)
Comment (gast04) [reply selfdrive/sensord/sensors_qcom2.cc]: done

File: selfdrive/sensord/sensors/bmx055_accel.cc (hunk 14 above)
Comment (gast04) [reply selfdrive/sensord/sensors/bmx055_accel.cc]: fixed

File: selfdrive/sensord/sensors/lsm6ds3_accel.cc (hunk 15 above)
Comment (adeebshihadeh) [reply selfdrive/sensord/sensors/lsm6ds3_accel.cc]: yeah let's clean this up

**Pull request merged**